README.md
# grunt-broccoli
[![Build Status](https://travis-ci.org/quandl/grunt-broccoli.svg?branch=master)](https://travis-ci.org/quandl/grunt-broccoli)

[Broccoli](http://broccolijs.com/) is a performant and well maintained build tool that has been blessed by the Ember community. [Grunt](https://gruntjs.com/) is the task manager that we all know and love (you know you still do). They fill their respective niches well, and using them together makes a lot of sense.

This Grunt plugin lets you specify different Broccoli tasks in the same Gruntfile, and then use either the `build`, `watch` or `serve` command with each of them.

## Upgrading from 0.4.2

The latest version of grunt-broccoli upgrades the Broccoli dependency to ^1.0.0. All previously existing options still remain, but to align better with Grunt convention, they are now specified in the `options` object of your task. This also enables you to specify global options for all Broccoli tasks that can then be overriden for individual tasks.

For more details, see the [changelog](https://github.com/EmberSherpa/grunt-broccoli/blob/master/CHANGELOG.md).

## Get started

Install it by running `npm i -D grunt-broccoli`.

We've included a sample Gruntfile configuration below. You can use any name for your targets and then, in the terminal, pass either `build`, `watch` or `serve` as the final command. For instance:

```sh
grunt broccoli:prod:build
# or
grunt broccoli:dev:watch
# or
grunt broccoli:inline-config:serve
```

```js
module.exports = function(grunt) {
  grunt.loadNpmTasks('grunt-broccoli');

  grunt.initConfig({
    broccoli: {
      dev: {
        dest: 'public/assets',
        // These are the default option values. You only have to specify the
        // properties that you want to change
        options: {
          // By default, grunt-broccoli will walk up the directory tree until it
          // finds a file named "Brocfile.js". But if you specify a custom
          // path, it will be used "as is". It's also possible to specify a
          // function that returns a Broccoli node if you prefer to keep your
          // whole build config in Gruntfile.js
          config: 'Brocfile.js',
          // Can be either a string or an object. If a string, it specifies the
          // value of process.env.BROCCOLI_ENV. If an object, each key and value
          // will be merged with process.env
          env: 'development',
          // Which host to use with the 'serve' task
          host: '127.0.0.1',
          // If true, grunt-broccoli will only overwrite the files that
          // actually have changed after every rebuild. If false, the 'dest'
          // directory will be removed completely, and then rewritten, after
          // every rebuild
          incrementalOverwrite: true,
          // Which port to use with the 'serve' task
          port: 4200,
          // This option dictates where temporary files are placed while
          // Broccoli executes a build. Before Broccoli 1.0.0, they were placed
          // in a directory called `tmp` next to the Brocfile. After 1.0.0,
          // they're normally placed in the OS's global tmp directory
          tmpdir: undefined
        }
      },
      prod: {
        dest: 'public/assets',
        options: {
          env: {
            BROCCOLI_ENV: 'production'
          }
        }
      },
      'inline-config': {
        dest: 'public/assets',
        options: {
          // Here's an example of an inlined config. It's identical to what you
          // would put in a Brocfile, aside from the fact that a Brocfile would
          // use `module.exports` instead of `return` for the output node
          config: function() {
            var Sass = require('broccoli-sass');
            var PostCSS = require('broccoli-postcss');
            var Funnel = require('broccoli-funnel');
            var MergeTrees = require('broccoli-merge-trees');

            var css = new Sass(['src/scss'], 'site.scss', 'css/site.css');

            css = new PostCSS(css, {
              plugins: [
                {
                  module: require('autoprefixer'),
                  options: { browsers: ['defaults', 'last 4 versions', 'IE 9'] }
                }
              ]
            });

            var img = new Funnel('src/img', { destDir: 'img' });

            return new MergeTrees([css, img]);
          }
        }
      }
    }
  });

  grunt.registerTask('build', ['broccoli:prod:build']);
  grunt.registerTask('watch', ['broccoli:dev:watch']);
  grunt.registerTask('serve', ['broccoli:dev:serve']);

  grunt.registerTask('default', ['watch']);
};
```

For a more exhaustive example of what a `Brocfile.js` might look like, see [broccoli-sample-app](https://github.com/broccolijs/broccoli-sample-app).


CHANGELOG.md
## 1.1.4
- When using `options.incrementalOverwrite`, always overwrite target files on first successful build
- Updated devDependencies to use `broccoli-sass-source-maps` and `sass` rather than the practically abandoned `broccoli-sass` package
- Updated test suite to also test watch task

## 1.1.3
- Fixed a bug where watch command wouldn't write anything if dest directory didn't already exist

## 1.1.2
- Updated findup-sync dependency to ^2.0.0

## 1.1.1

- Fixed a bug in `options.incrementalOverwrite` that would prevent items that previously hadn't been included in the build tree from being written.
- Fixed a bug where a single rebuild would be logged multiple times when `options.incrementalOverwrite` was enabled

## 1.1.0

Added `options.tmpdir`. Primarily useful if you depend on a Broccoli plugin that still assumes temporary build files to be placed in a `tmp` directory next to the Brocfile.

## 1.0.0

A complete rewrite with inspiration and code from [broccoli](https://www.npmjs.com/package/broccoli)'s own CLI module.

Broccoli ^1.0.0 is now used, and tests load Grunt ^1.0.0.

#### API changes

Please note that apart from `dest`, all options now live in the `options` object. This is to align more closely with Grunt convention, and enables you to specify global options that can be locally overridden. Just like with any other Grunt plugin!

```js
// Old config
grunt.initConfig({
  broccoli: {
    dev: {
      env 'production',
      dest: 'build/assets'
    }
  }
});

// New config
grunt.initConfig({
  broccoli: {
    dev: {
      dest: 'build/assets',
      options: {
        env 'production'
      }
    }
  }
});
```

#### New features

##### Incremental overwrite

The `options.incrementalOverwrite` property has been added! Previously, the plugin would mimic `broccoli-cli` in that the `dest` directory would be completely removed, then completely rewritten with every incremental rebuild. This works surprisingly well from a performance standpoint, but when integrating with other tools that watch the output directory, it can become overly noisy.

**Please note!** This option is turned on by default for `watch` and `serve` tasks.

##### Option validation

The task configuration is now validated against a schema. This should hopefully make it easier to get the configuration right.


test/test.js
var fs = require('fs');
var path = require('path');
var rimraf = require('rimraf');
var spawn = require('child_process').spawn;
var walkSync = require('walk-sync');

var initialSiteScss = fs.readFileSync(
  path.join(__dirname, 'fixtures', 'src', 'scss', 'site.scss'),
  {
    encoding: 'utf8'
  }
);

function assertBuildDir(buildDir, test, cssFilename, callback) {
  var destItems = walkSync(buildDir);

  test.deepEqual(destItems, [
    'css/',
    'css/site.css',
    'img/',
    'img/USDA_Broccolini.jpg',
    'index.html'
  ]);

  var builtCSS = fs.readFileSync(path.join(buildDir, 'css', 'site.css'), {
    encoding: 'utf8'
  });

  var fixtureCSS = fs.readFileSync(
    path.join(__dirname, 'fixtures', 'expected', cssFilename),
    {
      encoding: 'utf8'
    }
  );

  test.equal(builtCSS.trim(), fixtureCSS.trim());
  callback();
}

function runBuildTask(taskName) {
  return function(test) {
    test.expect(2);

    var buildDir = path.join(__dirname, 'build', taskName);
    rimraf.sync(buildDir);

    var child = spawn('grunt', ['broccoli:' + taskName + ':build']);

    child.on('close', function() {
      assertBuildDir(buildDir, test, 'initial.css', test.done);
    });
  };
}

// This test only tests that the watch task is able to trigger an initial build
// and that it rebuilds files after they have changed. It does not yet test when
// files are added or removed.
function runWatchTask(taskName) {
  return function(test) {
    test.expect(5);

    var buildDir = path.join(__dirname, 'build', taskName);
    rimraf.sync(buildDir);

    var child = spawn('grunt', ['broccoli:' + taskName + ':watch']);

    // Wait for the initial build
    child.stdout.on('data', function(data) {
      if (/Built \(\d+ items? changed\)/i.test(data.toString())) {
        child.stdout.removeAllListeners('data');

        // Assert that the initial build matches the usual build output
        assertBuildDir(buildDir, test, 'initial.css', function() {
          // Wait for a second build
          child.stdout.on('data', function(data) {
            // This method of checking the stdout output is a very crude way of
            // testing the options.incrementalOverwrite feature, but at least
            // it's something
            test.ok(/Built \(1 item changed\)/i.test(data.toString()));

            assertBuildDir(buildDir, test, 'second.css', function() {
              // Revert fixtures/src/scss/site.scss to its original state
              fs.writeFileSync(
                path.join(__dirname, 'fixtures', 'src', 'scss', 'site.scss'),
                initialSiteScss
              );

              child.kill('SIGINT');
              test.done();
            });
          });

          // Since options.incrementalOverwrite rounds file mtimes to the whole
          // seconds, we need to wait a little before triggering a file change
          // to site.scss
          setTimeout(function() {
            // Trigger a second build by appending a line to
            // fixtures/src/scss/site.scss
            fs.appendFileSync(
              path.join(__dirname, 'fixtures', 'src', 'scss', 'site.scss'),
              'a { text-decoration: none; }'
            );
          }, 1000);
        });
      }
    });
  };
}

exports.buildBrocfile = runBuildTask('brocfile');
exports.buildDefault = runBuildTask('default');
exports.buildFunction = runBuildTask('function');

exports.watchBrocfile = runWatchTask('brocfile');
exports.watchDefault = runWatchTask('default');
exports.watchFunction = runWatchTask('function');


test/fixtures/Brocfile.js
var Sass = require('broccoli-sass-source-maps')(require('sass'));
var PostCSS = require('broccoli-postcss');
var Funnel = require('broccoli-funnel');
var MergeTrees = require('broccoli-merge-trees');

var css = new Sass(['src/scss'], 'site.scss', 'css/site.css');

css = new PostCSS(css, {
  plugins: [
    {
      module: require('autoprefixer'),
      options: { browsers: ['IE 10'] }
    }
  ]
});

var img = new Funnel('src/img', { destDir: 'img' });

var static = new Funnel('src', {
  exclude: ['img', 'scss']
});

module.exports = new MergeTrees([css, img, static]);


Gruntfile.js
module.exports = function(grunt) {
  grunt.loadNpmTasks('grunt-contrib-nodeunit');
  grunt.loadNpmTasks('grunt-eslint');
  grunt.loadTasks('tasks');
  grunt.registerTask('test', ['eslint', 'nodeunit']);

  grunt.initConfig({
    eslint: {
      target: ['*.js', 'tasks/*.js', 'test/*.js']
    },
    nodeunit: {
      tests: ['test/*.js']
    },
    broccoli: {
      brocfile: {
        dest: 'test/build/brocfile',
        options: {
          config: './test/fixtures/Brocfile.js',
          tmpdir: 'tmp'
        }
      },
      function: {
        dest: 'test/build/function',
        options: {
          config: function() {
            var path = require('path');

            process.chdir(path.resolve(__dirname, 'test/fixtures'));
            return require('./test/fixtures/Brocfile.js');
          }
        }
      },
      default: {
        dest: 'test/build/default'
      }
    }
  });
};


lib/plugin.js
var path = require('path');
var broccoli = require('broccoli');
var rimraf = require('rimraf');
var mkdirp = require('mkdirp');
var path = require('path');
var copyDereferenceSync = require('copy-dereference').sync;
var Watcher = require('broccoli/lib/watcher');

var plugin = {
  builder: function(config) {
    var tree;
    if (typeof config === 'function') {
      tree = config();
    } else if (typeof config === 'string' || typeof config === 'undefined') {
      var configFile = config || 'Brocfile.js';
      var configPath = path.join(process.cwd(), configFile);
      try {
        tree = require(configPath);
      } catch(e) {
        // grunt.fatal("Unable to load Broccoli config file: " + e.message);
      }
    }
    return new broccoli.Builder(tree);
  },
  build: function(dest, config) {
    var builder = this.builder(config);
    return builder.build()
      .then(function(output) {
        rimraf.sync(dest);
        mkdirp.sync(path.join(dest, '..'));
        copyDereferenceSync(builder.outputPath, dest);
        return output;
      });
  },
  serve: function(config, options) {
    var builder = this.builder(config);
    broccoli.server.serve(builder, { host: options.host, port: options.port, liveReloadPort: options.liveReloadPort });
  },
  watch: function(dest, config) {
    var builder = this.builder(config);
    var watcher = new Watcher(builder, { interval: 100 });
    return watcher.on('change', function(results) {
      console.log('\n\nChange detected');
      rimraf.sync(dest);
      mkdirp.sync(path.join(dest, '..'));
      copyDereferenceSync(results.directory, dest);
      console.log('Build successful\n');
      watcher.emit("livereload");
    });
  }
};

module.exports = plugin;


tasks/broccoli.js
var _ = require('lodash');
var broccoli = require('broccoli');
var copyDereferenceSync = require('copy-dereference').sync;
var findup = require('findup-sync');
var fs = require('fs');
var Joi = require('joi');
var mkdirp = require('mkdirp');
var path = require('path');
var rimraf = require('rimraf');
var walkSync = require('walk-sync');

module.exports = function(grunt) {
  grunt.registerMultiTask(
    'broccoli',
    'Execute custom Broccoli task',
    broccoliTask
  );

  function broccoliTask() {
    var options = this.options({
      config: findup('Brocfile.js', { nocase: true }),
      env: 'development',
      host: '127.0.0.1',
      incrementalOverwrite: true,
      port: 4200
    });

    var command = this.args[0];

    var dataSchema = Joi.object().keys({
      dest: Joi.string(),
      options: Joi.object().optional()
    });

    var optionsSchema = Joi.object().keys({
      config: [Joi.string(), Joi.func()],
      env: ['development', 'production', Joi.object()],
      host: Joi.string().hostname(),
      incrementalOverwrite: Joi.boolean(),
      port: Joi.number(),
      tmpdir: Joi.string().optional()
    });

    Joi.attempt(
      this.data,
      dataSchema,
      'If you recently upgraded grunt-broccoli to ^1.0.0, please ensure that you moved all options into the options object.\n'
    );

    Joi.attempt(options, optionsSchema);

    Joi.attempt(
      command,
      ['build', 'watch', 'serve'],
      'grunt-broccoli supports the following commands: "build", "watch" and "serve".\n'
    );

    if (_.isString(options.env)) {
      process.env.BROCCOLI_ENV = options.env;
    } else if (_.isPlainObject(options.env)) {
      _.forEach(options.env, function(value, key) {
        process.env[key] = value;
      });
    }

    if (options.tmpdir) {
      options.tmpdir = path.resolve(process.cwd(), options.tmpdir);
      mkdirp.sync(options.tmpdir);
    }

    var originalCwd = process.cwd();
    var dest = path.resolve(process.cwd(), this.data.dest);
    var done = this.async();
    var outputNode;

    if (_.isString(options.config)) {
      var configPath = path.resolve(process.cwd(), options.config);
      process.chdir(path.dirname(configPath));

      outputNode = require(configPath);
    } else if (_.isFunction(options.config)) {
      outputNode = options.config();
    }

    var builder = new broccoli.Builder(outputNode, { tmpdir: options.tmpdir });

    if (command === 'build') {
      builder
        .build()
        .then(function() {
          var logMessage = 'Build successful';

          if (builder.outputNodeWrapper.buildState.totalTime) {
            var buildTime = Math.round(
              builder.outputNodeWrapper.buildState.totalTime
            );
            logMessage += ' (' + buildTime + 'ms)';
          }

          rimraf.sync(dest);
          mkdirp.sync(path.join(dest, '..'));
          copyDereferenceSync(builder.outputPath, dest);

          grunt.log.ok(logMessage);
          done();
        })
        .finally(function() {
          process.chdir(originalCwd);

          return builder.cleanup();
        })
        .catch(function(error) {
          if (error.file) {
            grunt.log.error('File: ' + error.file);
          }

          grunt.log.error(error.stack);
          grunt.log.error('\nBuild failed');

          done(false);
        });

      return;
    }

    // From here on out, we can assume that command is either "watch" or "serve"
    var watcher = new broccoli.Watcher(builder);
    var hasBuiltOnce = false;
    var changedItems = [];

    function cleanupAndExit() {
      return watcher.quit();
    }

    process.on('SIGINT', cleanupAndExit);
    process.on('SIGTERM', cleanupAndExit);

    watcher.on('buildSuccess', function() {
      var buildTime = Math.round(
        builder.outputNodeWrapper.buildState.totalTime
      );

      if (options.incrementalOverwrite) {
        mkdirp.sync(dest);

        var srcItems = walkSync(builder.outputPath);
        var destItems = walkSync(dest);

        var removedDestItems = _.difference(destItems, srcItems);

        removedDestItems.forEach(function(item) {
          var itemDestPath = path.join(dest, item);
          rimraf.sync(itemDestPath);
        });

        changedItems = removedDestItems.slice(0);

        srcItems.forEach(function(item) {
          var destMtime;

          var itemSrcPath = path.join(builder.outputPath, item);
          var srcStats = fs.statSync(itemSrcPath);
          var srcMtime = 1000 * Math.floor(srcStats.mtime.getTime() / 1000);

          try {
            var itemDestPath = path.join(dest, item);
            var destStats = fs.statSync(itemDestPath);
            destMtime = 1000 * Math.floor(destStats.mtime.getTime() / 1000);

            // The mtime of a directory is updated when the mtime of one of its
            // contained items are updated. By not copying whole directories, we
            // avoid overwriting other contained items that haven't changed
            if (srcStats.isDirectory() && destStats.isDirectory()) {
              return;
            }
          } catch (error) {
            destMtime = Number.NEGATIVE_INFINITY;
          }

          if (srcMtime > destMtime || !hasBuiltOnce) {
            changedItems.push(item);
            rimraf.sync(itemDestPath);
            mkdirp.sync(path.dirname(itemDestPath));

            copyDereferenceSync(itemSrcPath, itemDestPath);
          }
        });

        var changedItemsLog =
          changedItems.length +
          (changedItems.length === 1 ? ' item' : ' items') +
          ' changed';

        grunt.log.ok(
          'Built (' +
            changedItemsLog +
            ') - ' +
            buildTime +
            ' ms @ ' +
            new Date().toString()
        );
      } else {
        rimraf.sync(dest);
        mkdirp.sync(dest);

        copyDereferenceSync(builder.outputPath, dest);

        grunt.log.ok('Built - ' + buildTime + ' ms @ ' + new Date().toString());
      }

      hasBuiltOnce = true;
    });

    watcher.on('buildFailure', function(error) {
      grunt.log.error('Built with error:');
      grunt.log.error(error.message);

      if (!error.broccoliPayload || !error.broccoliPayload.location.file) {
        grunt.log.error(error.stack);
      }
    });

    var watcherPromise = watcher.start();

    watcherPromise
      .catch(function(error) {
        grunt.log.error((error && error.stack) || error);
      })
      .finally(function() {
        builder.cleanup();
      })
      .catch(function(error) {
        grunt.log.error('Cleanup error:');
        grunt.log.error((error && error.stack) || error);
      })
      .finally(function() {
        process.exit(1);
      });

    if (command === 'serve') {
      var connect = require('connect');
      var http = require('http');
      var middleware = require('broccoli/lib/middleware');
      var tinylr = require('tiny-lr');

      var tinylrServer = new tinylr.Server();
      tinylrServer.listen();

      var app = connect().use(middleware(watcher));
      var server = http.createServer(app);
      server.listen(options.port, options.host);

      watcher.on('buildSuccess', function() {
        tinylrServer.changed({ body: { files: changedItems } });
      });

      watcher.on('buildFailure', function() {
        tinylrServer.changed({ body: { files: [] } });
      });

      watcherPromise.finally(function() {
        tinylrServer.close();
        server.close();
      });
    }
  }
};
